Rework the front end preview

richardhj opened the pull request:
This PR injects the back end preview bar and replaces the preview frameset.

**How this works:**

The preview bar is injected into the DOM on `kernel.response` event.
When the user wants to view the page as a member, an Ajax request is triggered in which the preview authentication will be handled. The Ajax endpoint logs in the user, returns an HTTP OK  and finally, a page reload is done.

**TODO:**
- [x] CSS (rudimental CSS at the moment)
- [x] make the preview bar collapsible
- [x] restore contao/core#6546

**Impressions:**
<img width="1411" alt="Screen Shot 2019-11-24 at 16 09 40" src="https://user-images.githubusercontent.com/1284725/69491274-38fb3980-0ed6-11ea-97b6-625b730c9505.png">

____

**Outlook:**
I already worked on an enhanced version of the preview. Once this PR is approved and merged, I add a framework for 3rd parties to add content to the preview toolbar.
So-called "preview providers" (class must implementing an interface and tag the service with `contao.preview_provider`) can add arbitrary content to the toolbar.
In the end, the great advantage of the preview is the back end scope in the front end.
The toolbar should be either horizontally (as we know it) or left-aligned when the content requires it.

Use case: theme switch or changing CSS variables on runtime.

<img width="1509" alt="Screen Shot 2019-11-24 at 20 52 19" src="https://user-images.githubusercontent.com/1284725/69494228-57742b80-0efc-11ea-9254-0a1848112b65.png">

Toflar commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,161 @@
+
+        // Only inject toolbar on html responses
+        if ('html' !== $request->getRequestFormat()
+            || false === strpos((string) $response->headers->get('Content-Type'), 'html')
```
We check for `text/html` everywhere else :)

richardhj replied:
So can be removed safely?

Toflar replied:
No it should be adjusted to test for `text/html`, not just `html`. That's what I meant.
Just to be consistent with the rest of the source :)

Toflar commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,161 @@
+        $frontendUsername = $this->tokenChecker->getFrontendUsername();
+        $showUnpublished  = $this->tokenChecker->isPreviewMode();
+
+        $toolbar = str_replace(
```
Why this `str_replace()`?

richardhj replied:
This removes line-breaks and reduces the injected html,js,css to one line. If you view the page source, it is much more tidy. Symfony Debug Toolbar does this as well :)

Toflar reviewed (commented):
Awesome!! Really looking forward to this!

richardhj commented on core-bundle/src/Resources/contao/controllers/BackendPreview.php:
```diff
@@ -1,121 +0,0 @@
-
-		$strUrl = null;
-
-		if (Input::get('url'))
```
Needs to be re-added (by redirecting within the `PreviewToolbarListener`)

aschempp commented on core-bundle/src/Controller/BackendController.php:
```diff
@@ -107,7 +108,7 @@ public function previewAction(Request $request): Response
         $this->get('contao.framework')->initialize();
 
-        $controller = new BackendPreview();
+        $controller = new FrontendIndex();
```
Is this just for BC reason? Then we should make sure the route is no longer used and add a deprecation warning?

richardhj replied:
I addressed this

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,149 @@
+ *
+ * @Route(defaults={"_scope" = "backend"})
+ */
+final class BackendPreviewSwitchController
```
I don't think we use `final` (yet)

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,149 @@
+        $this->contaoFramework              = $contaoFramework;
+        $this->frontendPreviewAuthenticator = $frontendPreviewAuthenticator;
+        $this->tokenChecker                 = $tokenChecker;
+        $this->connection                   = $connection;
```
Do not align the equal operator

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    /**
+     * @Route("/_contao/preview_switch", name="contao_backend_preview_switch")
```
I think the `_contao/` routes are used for the front end only. Maybe we should use `/contao/preview_switch`? That should also make sure the security access control checks for the back end user.

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,149 @@
+    {
+        $this->contaoFramework->initialize(false);
+
+        $user = BackendUser::getInstance();
```
Should use the `security.helper` service to get the user.

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,149 @@
+
+    private function isAllowedToAccessMembers(BackendUser $user): bool
+    {
+        return ($user->isAdmin || (!empty($user->amg) && \is_array($user->amg)));
```
Should use security voters (at some point). Maybe we can even come up with a role for that?

richardhj replied:
There is `$this->security->isGranted('contao_user.amg' /* , $subject = null */,);` available but atm you MUST pass an array of member groups as second parameter. Therefore, we cannot use the existing `BackendAccessVoter` to check if a user has any member groups allowed.

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        // Ignore redirects and errors
+        if (200 !== $response->getStatusCode()) {
```
Maybe this should use `$response->isSuccessful()` or at least `$response->isOk()`

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,203 @@
+            return;
+        }
+
+        if (null === $user = BackendUser::getInstance()) {
```
Use the `security.helper` service

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,203 @@
+            "\n",
+            '',
+            $this->twig->render(
+                '@ContaoCore/Backend/preview_toolbar.html.twig',
```
Why is the preview toolbar in the `Backend` namespace?

aschempp reviewed (commented):
Very nice! Can we maybe add an image to see what the toolbar will look like?

Any particular reason why the user list is loaded with ajax?

xchs commented on core-bundle/src/Resources/views/FrontendPreview/authenticator.html.twig:
```diff
@@ -0,0 +1,154 @@
+    <style>
+        .cto-toolbar__section--authenticator {
+
+        }
```
Is this CSS rule necessary if there are no declarations?

richardhj replied:
These changes are postponed to a later PR now :)

richardhj committed:
Rework frontend preview

Toflar reviewed (commented):
Looks excellent to me. Imho, this can be extended with unit tests and merged as is!

Polishing CSS and adding more toolbar functionality such as JS can still be done afterwards. At least we got rid of the iframe now and the preview bar is shown consistently everywhere :) Great work!

richardhj committed:
Collapse toolbar

richardhj committed:
Check for frontend request

ausi commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,109 @@
+        );
+
+        $content = preg_replace(
+            '/<body[\\s\\S]*?>/',
```
```suggestion
            '/<body.*?>/is',
```

ausi commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        $content = $response->getContent();
+
+        if (false === strpos($content, '<body')) {
```
```suggestion
        if (false === stripos($content, '<body')) {
```

richardhj committed:
Apply suggestions from code review

Co-Authored-By: Martin Auswöger <martin@auswoeger.com>

richardhj committed:
Remove the contao_backend_switch route

richardhj committed:
Merge remote-tracking branch 'richardhj/feature/preview' into feature/preview

aschempp commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,118 @@
+            throw new RedirectResponseException($this->previewScript.$request->getRequestUri());
+        }
+
+        $this->contaoFramework->initialize(false);
```
You should not specify the `false` argument, that was only added to explicitly initialize when doing front end routing.

aschempp commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,118 @@
+
+        $this->contaoFramework->initialize(false);
+
+        if (!$this->authorizationChecker->isGranted('ROLE_USER')) {
```
Maybe extend `AbstractController` and use `$this->denyAccessUnlessGranted()`? Not sure if we have a decision on whether to use that though @contao/developers ?

Toflar replied:
I like DI better, always :)

richardhj replied:
Did you read https://github.com/contao/contao/pull/989#discussion_r350001024? we still need to create a Role, or voter.
Maybe `$this->security->isGranted('ROLE_ACCESS_MEMBER_GROUP', 3);`?

aschempp commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,118 @@
+            $this->frontendPreviewAuthenticator->removeFrontendAuthentication();
+        }
+
+        if ($request->query->get('url')) {
```
what is this parameter used for?

richardhj replied:
Maybe @leofeyer knows what's the original use case of this parameter. It definitely is obsolete (in contrast to the `page` parameter).
https://github.com/contao/contao/blob/1c1a157a2b497a874bbf67de3b9b37e14b63a284/core-bundle/src/Resources/contao/controllers/BackendPreview.php#L95-L98

leofeyer replied:
Why should the parameter be obsolete?

```
// Preview a certain page
https://domain.tld/contao/preview?page=5

// Preview a certain URL
https://domain.tld/contao/preview?url=en/content-elements.html
```

See https://github.com/contao/core/issues/6471

richardhj replied:
If you want to preview a certain URL, you don't need the `/contao/preview` route. You can just point to the target URL, e.g. `/preview.php/en/content-elements.html`.
Using the `url` parameter is only reasonable if you simultaneously need to authenticate a user and thus also pass the `user` parameter, e.g. `https://domain.tld/contao/preview?url=en/content-elements.html&user=j.doe`
So we should keep the parameter.

aschempp commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,118 @@
+            throw new RedirectResponseException($targetUrl);
+        }
+
+        if ($request->query->get('page') && null !== $page = PageModel::findWithDetails($request->query->get('page'))) {
```
Can we please move this to an event listener (same as anyone else would implement it)?

richardhj replied:
The `PreviewUrlConvertEvent`? Should I add the `$request` attribute to the constructor arguments of the `PreviewUrlConvertEvent` then?

richardhj replied:
Done in 94348967b6c2ebb3927c72575545139762729f11.

aschempp commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,118 @@
+        $this->dispatcher->dispatch($urlConvertEvent);
+
+        if (null !== $targetUrl = $urlConvertEvent->getUrl()) {
+            throw new RedirectResponseException($targetUrl);
```
This is a controller, you don't need to throw an exception, just return a redirect response 😆

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+     */
+    public function __invoke(Request $request): Response
+    {
+        $this->contaoFramework->initialize(false);
```
Please remove the argument

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+        $user = $this->security->getUser();
+
+        if (!($user instanceof BackendUser) || !$request->isXmlHttpRequest()) {
+            throw new PageNotFoundException('Bad response');
```
Why is that a `PageNotFound`? Again, this is a controller, maybe return a "Bad Request" response?

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+            try {
+                $toolbar = $this->renderToolbar($user);
+            } catch (TwigError $e) {
+                return Response::create('', Response::HTTP_INTERNAL_SERVER_ERROR);
```
I have never used `Response::create()` but just `new Response()`, not sure if that method is just a "leftover" from before-PHP 5.5?

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+        if ('tl_switch' === $request->request->get('FORM_SUBMIT')) {
+            $this->authenticatePreview($user, $request);
+
+            return Response::create();
```
If the response is empty, it should be `204` imho.

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+            return JsonResponse::create($data);
+        }
+
+        return Response::create('', Response::HTTP_INTERNAL_SERVER_ERROR);
```
This should be a "Bad Request" not a server error, right?

aschempp commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,206 @@
+     */
+    private function renderToolbar(BackendUser $user): string
+    {
+        $canSwitchUser = ($user->isAdmin || (!empty($user->amg) && \is_array($user->amg)));
```
Should use `$this->isAllowedToAccessMembers()` or even better we should add a user role for this so we can use security to decide

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,109 @@
+        $request = $event->getRequest();
+        $response = $event->getResponse();
+
+        if ($request->getScriptName() !== $this->previewScript) {
```
Is this the only way to detect the preview script? Be aware that the script can be empty, and I would assume then we would not inject anything?

richardhj replied:
What do you mean by "the script can be empty" and what do you propose?

Toflar replied:
The preview script is configurable so in theory it would be possible to enter `''` there. However, this would require quite a bit of reverse proxy configuration (if it's not impossible at all) and I don't think it's a use case that really should be supported. Configuring it to `p.php` or whatever, yes. But leaving it empty...I don't know. I'm fairly confident this will never happen.

aschempp replied:
it will happen, if you do not have a reverse proxy. Then you also do not need an entry point script…

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        $content = $response->getContent();
+
+        if (false === stripos($content, '<body')) {
```
```suggestion
        if (false === stripos($content, '<body ')) {
```

aschempp commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,109 @@
+            ]
+        );
+
+        $content = preg_replace(
```
Why are you adding this on the opening tag and not on the closing `</body>`?

richardhj replied:
Because I want the toolbar to be the first element within the `<body>` to be able to use `position:sticky`. Do you like b04c925b058082e388cacd797f19f90ea26ecd06 more?

aschempp replied:
🤔 I'm surprised now, I would say you can use sticky wherever you want in the dom. Also, this only injects the javascript, not the final DOM element?

richardhj replied:
position:sticky is a hybrid of static and fixed position, so the DOM element has to be the top one in our case.

> Also, this only injects the javascript, not the final DOM element?

I modified the template to only include the JS. That JS will put inject the toolbar wrapping `<div>`.

richardhj committed:
Apply suggestions from code review

richardhj committed:
Inject js before </body>

richardhj committed:
Remove non-present route

richardhj committed:
Move logic to PreviewUrlConvertEvent

richardhj committed:
Add tests

Toflar commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,205 @@
+            try {
+                $toolbar = $this->renderToolbar($user);
+            } catch (TwigError $e) {
+                return new Response('', Response::HTTP_INTERNAL_SERVER_ERROR);
```
I think I would force `renderToolbar()` to return a `string` and catch the error inside there and return a general `could not render twig template xy beacuse of: $e->getMessage()`. It shouldn't happen in production anyway as you're not adjusting templates there (especially not twig ones). So it's more of a message when adjusting the template in dev mode and I think in that case a proper error message would be helpful.

Toflar commented on core-bundle/tests/Controller/BackendPreviewControllerTest.php:
```diff
@@ -0,0 +1,172 @@
+        );
+
+        /** @var RedirectResponse $response */
+        $response = $controller->__invoke(new Request());
```
`$controller(new Request());`. Everywhere you used `->__invoke()` :)

Toflar reviewed (dismissed):
I left some minor nitpicks that could be improved for code readability and debugging.
There's one question that needs to be clarified with @contao/developers: Should we even allow empty `previewScript` (so the name for `preview.php`) configurations. If so, then this listener will have a hard time finding out if it needs to be invoked or not (and so do probably other listeners). We'd need to somehow check for the preview token in the session which is of course doable. It's just not necessary if we decide that an empty preview script is never supported. In theory it would be possible but it would require a huge amount of proxy configuration and re-implementing some kind of preflight-request thingy we had pre 4.8. I just don't think anyone is ever going to take that amount of work on their shoulders only to get rid of the `preview.php` in the URL. I do, however, like that it's configurable so one can go with `p.php` or whatever.

Other than that, this PR is in marvellous shape and should imho be reviewed and merged soon @leofeyer. Really great work @richardhj!

richardhj committed:
Apply suggestions from code review

richardhj commented:
I vote for NOT allowing empty string (`''`) nor `null` for the `%contao.preview_script%` parameter.

Note, that I found the simple comparison `$request->getScriptName() !== $this->previewScript` in other sources as well. Namely the `TokenChecker::isPreviewMode()` method.

aschempp commented:
> I vote for NOT allowing empty string (`''`) nor `null` for the `%contao.preview_script%` parameter.

An empty preview script is the case if you're running Contao in a regular Symfony application. One might not add a "secondary" entry point script. A preview script is also only necessary if you have a caching proxy, because we need it to "rewrite" the URLs. For an application that does not have a caching proxy, a secondary entry point is not necessary.
Be aware that this also means _using_ the entry point is not equal to _doing preview_! Or even more so, if the preview script is empty, there should probably not be a preview toolbar available.


> Note, that I found the simple comparison `$request->getScriptName() !== $this->previewScript` in other sources as well. Namely the `TokenChecker::isPreviewMode()` method.

That's the opposite of your comparison though, right?

richardhj commented:
> That's the opposite of your comparison though, right?

Exactly the same comparision:
IF `$request->getScriptName() !== $this->previewScript` -> `isPreviewMode()` = false
IF `$request->getScriptName() !== $this->previewScript` ->  Preview Toolbar does not get injected

Also this check is done in `PreviewAuthenticationListener::onKernelRequest()` BUT checking also for `'' === $this->previewScript`.

That means, we also need to fix the `TokenChecker::isPreviewMode()`.

> Be aware that this also means _using_ the entry point is not equal to _doing preview_! Or even more so, if the preview script is empty, there should probably not be a preview toolbar available.

Correct. 
But why would you then expect there is no preview toolbar when in preview mode but on the index.php entry point?

richardhj committed:
Do not inject preview bar if no preview entrypoint defined

richardhj commented:
Preview bar does not get injected if no preview entrypoint defined (since 2974a20)

aschempp reviewed (changes requested):
As discussed, please add a user role for member group access, otherwise this looks very good to me 😎

richardhj committed:
Add ROLE_ALLOWED_TO_SWITCH_MEMBER

Toflar reviewed (dismissed):
Great work, so much looking forward to this! <3

aschempp reviewed (changes requested):
I just installed this locally and noticed a few things:

1. entering a username lead to error-400 ajax requests (Safari/Mac). The Symfony Inspector shows them nicely
2. The preview bar is not responsive/mobile-friendly at all. I can't collapse it on a mobile either.
3. The preview bar does not remember its state, meaning it will re-open on every reload/navigation.
4. The Symfony bar is an overlay over the current page. Our preview bar will add a top margin. Not sure what's better?
5. The close icon will remain when the preview bar is closed. Maybe we can use the Contao icon similar to Symfony?

richardhj committed:
Further styling of toolbar

richardhj commented:
>     1. entering a username lead to error-400 ajax requests (Safari/Mac). The Symfony Inspector shows them nicely

I cannot reproduce. Can you provide the actual error and make sure the error is not due to an invalid cache?

I wanted to tackle the following points in a furhter PR, but here we go 😄 

>     2. The preview bar is not responsive/mobile-friendly at all. I can't collapse it on a mobile either.

Fixed.

>     3. The preview bar does not remember its state, meaning it will re-open on every reload/navigation.

Fixed utilizing local storage.

>     4. The Symfony bar is an overlay over the current page. Our preview bar will add a top margin. Not sure what's better?

The current approach does not add a top margin. More precisely the preview bar is prepended before the wrapper with relative/sticky positioning, thus the website will not get cropped.

I like this approach since a cropped website can be very annoying.

>     5. The close icon will remain when the preview bar is closed. Maybe we can use the Contao icon similar to Symfony?

Fixed. I attached some screenshots of the current version.
___

![collapse](https://user-images.githubusercontent.com/1284725/70292801-8eaeda80-1822-11ea-9c78-e07b320de6c2.gif)
<img width="393" alt="Screen Shot 2019-12-06 at 12 11 57" src="https://user-images.githubusercontent.com/1284725/70292813-9d958d00-1822-11ea-8158-7259bf8bd396.png">

richardhj committed:
Fix safari incompatibilites

richardhj commented:
I did not expect different behavior in Safari. I fixed two Safari incompatibilities (1x CSS, 1x JS) in 0732765, including the fix for the datalist @aschempp

aschempp commented:
For a minute, I thought about whether we can use [fetch](https://caniuse.com/#feat=fetch) instead of XmlHttpRequest. It does not work in IE11, but not sure if that's a problem.

**Then** I thought: why do we need Ajax at all? Submitting the preview form could be a simple POST request to the preview endpoint, including the return URL. We need to reload the page anyway after submitting? On the other hand, the whole preview does not work without JS anyway… it would just be a simpler implementation and not have ever-failing javascript 😆 WDYT?

richardhj commented:
Sure we can do that. Good idea! But only for the authentication. The datalist and initial fetching still rely on Ajax then.

richardhj commented:
> why do we need Ajax at all? Submitting the preview form could be a simple POST request to the preview endpoint, including the return URL. 

This suggests to split up the BackendSwitchController in more atomic parts. I would suggest to postpone these changes. Atm it is working fine :-)

richardhj committed:
Merge branch 'master' into feature/preview

richardhj committed:
Fix QS

Toflar reviewed (dismissed):
@leofeyer this is RTM imho. Please review :)

leofeyer commented on core-bundle/src/Event/PreviewUrlConvertEvent.php:
```diff
@@ -21,6 +22,13 @@ class PreviewUrlConvertEvent extends Event
 
+    private $request;
+
+    public function __construct(Request $request)
```
What is the advantage of storing the request in the event instead of injecting the request stack? Convenience?

Toflar replied:
It's just an integral part of that event. Just like it is for `kernel.request`, `kernel.response` etc.

leofeyer commented on core-bundle/src/Resources/contao/library/Contao/User.php:
```diff
@@ -561,6 +561,7 @@ public function serialize()
 			'username' => $this->username,
 			'password' => $this->password,
 			'admin' => $this->admin,
+			'amg' => $this->amg,
```
Why is this necessary? It will invalidate the token if the allowed member groups change.

richardhj replied:
If the "amg" property gets not serialized, the property is not present when checking/populating the token, and as a consequence, this will fail/never be the case: https://github.com/contao/contao/pull/989/commits/dfe0d2a984e6c3117bcc75a566bb0e13c96389de#diff-8b7196803029c7e93c87690d75f8ff91R590

leofeyer replied:
As discussed in Mumble on December 19th, this is correct, because Symfony does not allow to refresh roles on the fly.

leofeyer commented on core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig:
```diff
@@ -0,0 +1,277 @@
+        color: #666666;
+    }
+</style>
+<script>/*<![CDATA[*/
```
Why do we need `<![CDATA[` here? Isn't this required for XHTML only?

richardhj replied:
This is adopted from Symfony's toolbar listener. I guess Symfony has to support XHTML while Contao is not using XHTML anymore

richardhj replied:
CDATA removed in 206d929b.

leofeyer commented on core-bundle/src/Controller/BackendPreviewSwitchController.php:
```diff
@@ -0,0 +1,194 @@
+     */
+    public function __invoke(Request $request): Response
+    {
+        $this->contaoFramework->initialize();
```
Why is this necessary?

richardhj replied:
Removed in 2e6aa15a.

leofeyer committed:
Fix the coding style

leofeyer commented on core-bundle/src/Controller/BackendPreviewController.php:
```diff
@@ -0,0 +1,109 @@
+            return new RedirectResponse($this->previewScript.$request->getRequestUri());
+        }
+
+        $this->framework->initialize();
```
Is this really necessary?

richardhj replied:
Removed in 2e6aa15a.

leofeyer commented on core-bundle/src/Resources/config/listener.yml:
```diff
@@ -132,6 +132,23 @@ services:
             # The priority must be higher than the one of the Twig exception listener (defaults to -128)
             - { name: kernel.event_listener, event: kernel.exception, method: onKernelException, priority: -96 }
 
+    contao.listener.preview_bar:
```
The new listener need to be tested in the `ContaoCoreExtensionTest` class.

leofeyer commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -69,6 +69,31 @@ services:
             - '%kernel.project_dir%'
         public: true
 
+    Contao\CoreBundle\Controller\BackendPreviewController:
```
The new services need to be tested in the `ContaoCoreExtensionTest` class.

richardhj replied:
Done in 550a56124dd8073734a8b40d8a1ad1b232c38e31.

leofeyer commented on core-bundle/tests/Controller/BackendPreviewControllerTest.php:
```diff
@@ -0,0 +1,176 @@
+        $this->assertSame('preview.php', $response->getTargetUrl());
+    }
+
+    public function testThrowsAccessDenied(): void
```
Method names should be more "speaking": `testDeniesAccesIf…()`

leofeyer commented on core-bundle/tests/EventListener/PreviewToolbarListenerTest.php:
```diff
@@ -0,0 +1,345 @@
+    /**
+     * @dataProvider getInjectToolbarData
+     */
+    public function testInjectToolbar($content, $expected): void
```
This seems to be the same test as `testToolbarIsInjected()`?

richardhj replied:
One test is checking the bare snipped, i.e. if the snippet is correct, while the other test is checking whether the toolbar is injected on valid response.

leofeyer reviewed (changes requested):
This PR is looking pretty good. 👍 

I have commented what I would like to have discussed and/or changed. One thing I do not agree with is that we now invalidate the token if `$user->amg` changes. This does not make sense to me.

aschempp reviewed (changes requested):
As discussed in the Mumble call yesterday, `$user->amg` is the correct implementation, but it should be in the `BackendUser` class instead.

leofeyer commented:
> `$user->amg` is the correct implementation, but it should be in the `BackendUser` class instead.

#1120 has been merged, so this can be added now. 👍

richardhj committed:
Merge remote-tracking branch 'upstream/master' into feature/preview

# Conflicts:
#	calendar-bundle/tests/DependencyInjection/ContaoCalendarExtensionTest.php
#	calendar-bundle/tests/EventListener/PreviewUrlConverterListenerTest.php
#	core-bundle/src/Resources/contao/library/Contao/User.php
#	news-bundle/tests/DependencyInjection/ContaoNewsExtensionTest.php
#	news-bundle/tests/EventListener/PreviewUrlConverterListenerTest.php

richardhj committed:
Serialize amg in BackendUser.php

richardhj committed:
Remove CDATA as we use HTML

richardhj committed:
Do not initialize framework.

richardhj committed:
Test service registration.

richardhj commented:
I addressed the changes and merged against master to resolve conflicts.

aschempp reviewed (dismissed):
Looks all good to me! Unfortunately, the coverage seems to decrease quite a lot, are we missing tests on new classes?

leofeyer commented on core-bundle/src/EventListener/PreviewToolbarListener.php:
```diff
@@ -0,0 +1,115 @@
+        $this->router = $router;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
```
The method should be changed to `__invoke()` so we can simplify the event registration (see #1125).

leofeyer commented on core-bundle/src/EventListener/PreviewUrlConvertListener.php:
```diff
@@ -0,0 +1,68 @@
+        $this->framework = $framework;
+    }
+
+    public function onPreviewUrlConvert(PreviewUrlConvertEvent $event): void
```
The method should be changed to `__invoke()` so we can simplify the event registration (see #1125).

leofeyer committed:
Fix the coding style

leofeyer reviewed (changes requested):
The PR looks very good to me. 👍 We cannot merge it with the decreasing test coverage though.

leofeyer committed:
Fix the CSS code

Toflar commented on core-bundle/src/Resources/contao/library/Contao/Controller.php:
```diff
@@ -1142,7 +1142,6 @@ protected static function replaceOldBePaths($strContext)
 			'contao/password.php'  => $generate('contao_backend_password'),
 			'contao/popup.php'     => $generate('contao_backend_popup'),
 			'contao/preview.php'   => $generate('contao_backend_preview'),
-			'contao/switch.php'    => $generate('contao_backend_switch')
```
In this file there's still the `win.parent.postMessage` JS code which was used to communicate with the iframe previously. This can be removed as well now.

richardhj replied:
@leofeyer can you review 7666ac6aa0b10026d14f5fca830432ed0306e7c8?

Toflar replied:
This is correct 👍

richardhj committed:
Simplify event listener registration.

richardhj committed:
Remove legacy preview script code.

richardhj committed:
Add BackendPreviewSwitchControllerTest.php

leofeyer committed:
Adjust the test method names

leofeyer commented:
Thank you @richardhj. 🎉

Toflar commented:
🎉 thanks, Richard! ❤️
